frontend/src/modules/api/core.ts

```diff
@@ -0,0 +1,101 @@
+import { MIME_SUBTYPE, MIME_TYPE, RESPONSE_TYPE } from "./constant";
+
+interface IApiOptions {
+  method?: string;
+  headers?: {
+    "Content-Type"?: string;
+    Authorization?: string;
+  };
+  body?: string;
+}
+
+interface IApiResponse<T> {
+  statusCode: number;
+  data: T;
+}
+
+export const api = <T>(
+  url: string,
+  options?: IApiOptions
+): Promise<IApiResponse<T>> => {
+  const xhr = new XMLHttpRequest();
+  return new Promise((resolve, reject) => {
+    setXhrOptions(xhr, url, options);
+    xhr.onreadystatechange = function () {
+      if (xhr.readyState === XMLHttpRequest.DONE) {
+        const status = xhr.status;
+        const { type } = parseContentType(
+          xhr.getResponseHeader("Content-type")
+        );
+
+        isOk(status)
+          ? resolve(parseResponse<T>(type, xhr))
+          : reject({ statusCode: xhr.status, data: xhr.response });
```

Comment:
```suggestion
        const result = {
          statusCode: xhr.status,
          data: xhr.response,
        };
        if (xhr.responseType === RESPONSE_TYPE.JSON) {
          result.data = JSON.parse(xhr.response);
        }

        isOk(result.statusCode) ? resolve(result) : reject(result);
```

이런 방식은 어떨까요?

Reply:
```suggestion
        const statusCode = xhr.status;
        const data = isTypeJSON(xhr) ? JSON.parse(xhr.response) : xhr.response;
        const result = { statusCode, data };

        isOk(statusCode) ? resolve(result) : reject(result);
```

```ts
function isTypeJSON(xhr: XMLHttpRequest) {
  return xhr.response === RESPONSE_TYPE.JSON;
}
```
함수를 하나 분리하면 이렇게도 표현 가능하겠군요!